### DISCUSSION TITLE
Propper cleanup of raycasts?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#143] [Q&A] [answered]
I'm working on replicating a threeJS raycast demo which is fairly intense.
At the minimum level I'm doing 150 raycasts at 60fps.

After about a minute I hit a Memory assertion:
![Screenshot 2024-04-05 005542](https://github.com/jrouwe/JoltPhysics.js/assets/1397052/a594eb9e-b268-41f2-af68-ea494299ab94)

There's three spots I think could be issues for memory cleanup.

#### 1&2. : Setting the ray origin and direction. 
In every raycast call I update the ray origin and destination by setting the values with new vectors. 
I'm not sure what happens to the replaced Jolt Vectors. 

I do this because it's faster to use my threeJS > Jolt Vector handler.

I could rewrite my wrapper to set the vectors rather than replace them...

#### 3. `RaycastResults`
Do I need to destroy `RaycastResults`? I process them to create my own `hitResult` javascript object but haven't been destroying them. 

_(Sidenote I totally get that raycasting in the physics library like this isn't super efficient, but it's a good demo on using the raycasters)_

Is there anything else I should be looking out for in regards to raycasts?

**Comment by LeXXik** [comment]
If you are manually allocating memory for a new vector (e.g. `const jv = new Jolt.Vec3(0,0,0);`), then you need to manually free it (e.g. `Jolt.destroy(jv)`). You can refer to: https://github.com/jrouwe/JoltPhysics.js?tab=readme-ov-file#memory-management

What's a `RaycastResults`? Can you link to the line you are referring to in the example?

![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/d697d878-c781-4e3d-ba17-79d550b8153e)

**Reply by DennisSmolek** [reply]
`RaycastResult` is the typescript name for the response of the narrowphase query cast ray.
https://jrouwe.github.io/JoltPhysics/class_ray_cast_result.html

And I don’t declare them explicitly with additional references, do something like `ray.mOrigin = new Jolt.Vec3(0,0,0)`

**Reply by DennisSmolek** [reply]
Here is the raycast example where they are just replacing the origin and direction with new vectors:
https://github.com/jrouwe/JoltPhysics.js/blob/aa742fda357053bafdf807a043e573009f8833aa/Examples/ray_cast.html#L86

They do this in the example, although it’s stated that the examples are actually bad at memory management…

**Reply by jrouwe** [reply, marked as answer]
Yes, the construct above leaks memory. The proper way here would be:

```
let temp = new Jolt.RVec3(-5, 1, 4);
ray.mOrigin = temp;
Jolt.destroy(temp);
```

Or alternatively:

```
ray.mOrigin.Set(-5, 1, 4);
```

**Reply by jrouwe** [reply]
Note that the [build section](https://github.com/jrouwe/JoltPhysics.js?tab=readme-ov-file#building) also tells you how you can run with the memory profiler, which will tell you exactly where you're leaking.

**Reply by DennisSmolek** [reply]
What about when I call that value? Is it a reference or a new vector?
I have a helper that converts jolt vectors to three ones, ex:
```js
const vec3 {
    joltToThree: (vec: Jolt.Vec3, out = new THREE.Vector3()) =>
        out.set(vec.GetX(), vec.GetY(), vec.GetZ())
}
```
Then I call:
```js
const origin = vec3.joltToThree(ray.mOrigin);
```
By using `.GetX()` I would assume it's a reference, but it might be copying it behind the scenes.

**Reply by LeXXik** [reply]
There is no allocation for Jolt vector in this case. It uses existing vector. There is an allocation for THREE, but its on JS side, so GC will clean it, once no longer used. And `.GetX()` only reads a scalar value of the vector component, there is no allocation.

**Reply by DennisSmolek** [reply]
Something else to keep aware of is references to Jolt objects and nested values.

I was doing `this.bodyID = mHit.BodyID`
and then later doing `Jolt.destroy(mHit)`

my `this.bodyID` was a reference to the jolt object, so when I destroyed `mHit` it also destroyed the referenced `BodyID`
later when calling it, you'd expect an error, but nothing throws until you try and use the `BodyID`. I don't know what kind of object it was/is but it for sure wasn't the right bodyID.

**Reply by LeXXik** [reply]
Right. So, body can also become invalid in other cases, e.g. in raycast callbacks, which will receive a body ID on hit. That ID can point to a body that was already destroyed. You can always check if the body is actually valid with:

```js
if (Jolt.getPointer(body) === 0) {
  // body is no longer available
}
```

**Reply by jrouwe** [reply]
I don't think that last statement is true. It is perfectly possible that you store a `body` (not ID) somewhere and when it gets destroyed its pointer will still be what it was (even though it points to garbage now). `BodyLockInterface.TryGetBody` will allow you to check if a body ID is still valid.
